include/sparrow/buffer.hpp

```diff
@@ -686,6 +702,132 @@ namespace sparrow
         erase_at_end(get_data().p_begin);
     }
 
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, const T& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        return emplace(pos, value);
+    }
+
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, T&& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        return emplace(pos, std::move(value));
+    }
+
+    template <class T>
+    constexpr auto buffer<T>::insert(const_iterator pos, size_type count, const T& value) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        const difference_type __offset = pos - cbegin();
+        pointer ptr = get_data().p_begin + __offset;
+        if (count != 0)
+        {
+            const size_type sz = size();
+            const size_type off = ptr - get_data().p_begin;
+            if (const size_type new_size = sz + count; new_size > capacity())
+            {
+                reserve(new_size);
+            }
+            auto& data = get_data();
+            const auto new_ptr = data.p_begin + off;
+            std::move_backward(new_ptr, data.p_end, data.p_end + count);
+            std::fill_n(new_ptr, count, value);
+            data.p_end += count;
+        }
+        return begin() + __offset;
+    }
+
+    template <class T>
+    template <class InputIt>
+    constexpr auto buffer<T>::insert(const_iterator pos, InputIt first, InputIt last) -> iterator
+    {
+        SPARROW_ASSERT_TRUE(cbegin() <= pos && pos <= cend())
+        const size_type num_elements = last - first;
+        const size_type new_size = size() + num_elements;
+        const size_type offset = pos - cbegin();
+        reserve(new_size);
+        auto& data = get_data();
+        const pointer p = data.p_begin + offset;
+        std::move_backward(p, data.p_end, data.p_end + num_elements);
+        std::copy(first, last, p);
```

Comment:
Potential future optimization: a copy_or_move function which if the conditions are right (range moved in) will move instead of copy, otherwise copy. There was an implemetnation in Boost.Move but not sure if `std::copy` doesnt do that already under specific conditions.